### PR TITLE
wasm: capability restriction

### PR DESCRIPTION
**Pull request by ryanapilado**
Signed-off-by: Ryan Apilado <rapilado@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: wasm: implement capability restriction
Additional Description: Defines the configuration for capability restriction and propagates the configuration to `proxy-wasm-cpp-host`.
Risk Level: Medium
Testing: Unit tests ([`wasm_test.cc`](https://github.com/envoyproxy/envoy/pull/13911/files#diff-bc05b247c9a740af67a404a40498f96ced6bfc5911a2695f36bd60f0d348eadd)), configuration tests ([`config_test`](https://github.com/envoyproxy/envoy/pull/13911/files#diff-d743e0cfc1194f55ff4d936715cae872062265816c13d47c7db218cb9c3266ea)), and integration tests ([`wasm_filter_test.cc`](https://github.com/envoyproxy/envoy/pull/13911/files#diff-2ed9d95165b50294925d720defe225098c88b58381637c6457a0ea975a607d07)). Test that capabilities can be restricted selectively and are unrestricted by default.
Docs Changes: Add documentation for the new `messages` created in `wasm.proto`.
Release notes: N/A

### CODE REVIEW

**Comment by antoniovicente** [comment]
Thanks for the additional tests.  What is missing before this PR is ready for review?

File: test/extensions/filters/network/wasm/BUILD [hunk 1]
```diff
@@ -41,7 +41,6 @@ envoy_extension_cc_test(
     name = "wasm_filter_test",
     srcs = ["wasm_filter_test.cc"],
     data = envoy_select_wasm([
-        "//test/extensions/filters/network/wasm/test_data:logging_rust.wasm",
```
Comment (PiotrSikora) [inline test/extensions/filters/network/wasm/BUILD:44]: Unnecessary change?

File: test/extensions/filters/network/wasm/config_test.cc [hunk 2]
```diff
@@ -176,6 +176,93 @@ TEST_P(WasmNetworkFilterConfigTest, FilterConfigFailOpen) {
...
+  const std::string yaml = TestEnvironment::substitute(absl::StrCat(R"EOF(
+  config:
+    vm_config:
+      runtime: "envoy.wasm.runtime.)EOF",
+                                                                    GetParam(), R"EOF("
+      code:
+        local:
+          filename: "{{ test_rundir }}/test/extensions/filters/network/wasm/test_data/test_cpp.wasm"
+    capability_restriction_config:
+      allowed_abi_functions:
```
Comment (PiotrSikora) [inline test/extensions/filters/network/wasm/config_test.cc:192]: I think this needs to be `allowed_abi_functions: {}` for empty value.

File: test/extensions/common/wasm/wasm_test.cc [hunk 3]
```diff
@@ -933,6 +947,210 @@ TEST_P(WasmCommonTest, RemoteCodeMultipleRetry) {
...
+                    log_(spdlog::level::info, Eq("after on_vm_start, before proxy_log")))
+            .Times(0);
+        return root_context;
+      });
+  wasm->start(plugin);
+}
+
+// test with both proxy_on_vm_start and proxy_log allowed
+TEST_P(WasmCommonTest, AllowABIFunctions) {
+  if (GetParam() != "v8") {
```
Comment (PiotrSikora) [inline test/extensions/common/wasm/wasm_test.cc:1050]: Why are all those tests limited to V8?

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 4]
```diff
@@ -18,6 +18,12 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: Wasm]
 // [#extension: envoy.bootstrap.wasm]
 
+// Configuration for restricting proxy-wasm capabilities available to modules.
+message CapabilityRestrictionConfig {
+  // The list of proxy-wasm capabilities which will be exposed to the module.
+  repeated string allowed_abi_functions = 1;
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/wasm.proto:24]: Similarly to the proxy-wasm-cpp-host comment, I think this should be `allowe_capabilities` or similar.

**Review by PiotrSikora — COMMENTED** [review]
cc @mathetake @kyessenov

File: test/extensions/filters/network/wasm/BUILD (hunk 1 above)
Comment (ryanapilado) [reply test/extensions/filters/network/wasm/BUILD]: Fixed.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 4 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Fixed.

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 5]
```diff
@@ -18,6 +18,12 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: Wasm]
 // [#extension: envoy.bootstrap.wasm]
 
+// Configuration for restricting proxy-wasm capabilities available to modules.
+message CapabilityRestrictionConfig {
+  // The list of proxy-wasm capabilities which will be exposed to the module.
```
Comment (htuch) [inline api/envoy/extensions/wasm/v3/wasm.proto:23]: Can you add a reference to where the capabilities are defined?

File: test/extensions/filters/network/wasm/config_test.cc (hunk 2 above)
Comment (ryanapilado) [reply test/extensions/filters/network/wasm/config_test.cc]: Fixed.

File: test/extensions/common/wasm/wasm_test.cc (hunk 3 above)
Comment (ryanapilado) [reply test/extensions/common/wasm/wasm_test.cc]: Fixed.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 5 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Added a reference and updated the docs.

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 6]
```diff
@@ -18,6 +18,33 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  // and so should be left empty.
+  //
+  // The capability names are given in the
+  // `proxy-wasm ABI <https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT>`_.
+  // Additionally, the following WASI capabilities from
+  // `this list <https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#modules>`_
+  // are implemented and can be allowed:
+  // *fd_write*, *fd_read*, *fd_seek*, *fd_close*, *fd_fdstat_get*, *environ_get*, *environ_sizes_get*,
+  // *args_get*, *args_sizes_get*, *proc_exit*, *clock_time_get*, *random_get*.
+  map<string, SanitizationConfig> allowed_capabilities = 1;
```
Comment (mathetake) [inline api/envoy/extensions/wasm/v3/wasm.proto:34]: should this be `capability_configs`, or something like that? The content is not necessarily `allowlist` and sometimes `denylist`, isn't it

Comment (mathetake) [reply api/envoy/extensions/wasm/v3/wasm.proto:34]: same goes for proxy-wasm-cpp-host side PR

Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto:34]: The keys in the map are always allowed capabilities, so I think the name is still accurate. The `SanitizationConfig` which each capability maps to contains either an allowlist or denylist of arguments, but the capability itself is always allowed.

Comment (mathetake) [reply api/envoy/extensions/wasm/v3/wasm.proto:34]: ah, right. Thanks

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 7]
```diff
@@ -18,6 +18,33 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  // The proxy-wasm capabilities which will be allowed. Capabilities are mapped by
+  // name. The *SanitizationConfig* which each capability maps to is currently unimplemented and ignored,
+  // and so should be left empty.
+  //
+  // The capability names are given in the
+  // `proxy-wasm ABI <https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT>`_.
+  // Additionally, the following WASI capabilities from
+  // `this list <https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#modules>`_
+  // are implemented and can be allowed:
+  // *fd_write*, *fd_read*, *fd_seek*, *fd_close*, *fd_fdstat_get*, *environ_get*, *environ_sizes_get*,
```
Comment (mathetake) [inline api/envoy/extensions/wasm/v3/wasm.proto:32]: everything listed here except fd_write is not supported actually as you can see, for instance, https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/src/exports.cc#L727-L732 .

Also clock_time_get and get_random are implemented now https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/src/exports.cc#L804-L829

And in the future, we may support get_environ since there's a feature request in the multiple SDK repositories, https://github.com/proxy-wasm/spec/issues/19.

Comment (PiotrSikora) [reply api/envoy/extensions/wasm/v3/wasm.proto:32]: Even though the host implementation of WASI is indeed a bunch of mostly non-working (but adhering to spec) stubs, modules compiled against WASI want to import them, and this PR is purely about allowing/rejecting such calls.

File: test/extensions/filters/network/wasm/wasm_filter_test.cc [hunk 8]
```diff
@@ -190,6 +193,85 @@ TEST_P(WasmNetworkFilterTest, SegvFailOpen) {
...
+              log_(spdlog::level::trace, Eq(absl::string_view("onDownstreamConnectionClose 2 1"))))
+      .Times(0);
+  read_filter_callbacks_.connection_.close(Network::ConnectionCloseType::FlushWrite);
+  // Noop.
+  read_filter_callbacks_.connection_.close(Network::ConnectionCloseType::FlushWrite);
+  filter().testClose();
+}
+
+TEST_P(WasmNetworkFilterTest, RestrictOnDownstreamConnectionClose) {
+  if (std::get<0>(GetParam()) != "v8" || std::get<1>(GetParam()) != "cpp") {
```
Comment (mathetake) [inline test/extensions/filters/network/wasm/wasm_filter_test.cc:223]: I think this test should work with not only v8 && cpp but also other runtimes and languages except  'null' plugin, though I haven't testes.

File: test/extensions/filters/network/wasm/wasm_filter_test.cc [hunk 9]
```diff
@@ -190,6 +193,85 @@ TEST_P(WasmNetworkFilterTest, SegvFailOpen) {
...
+  EXPECT_CALL(filter(),
+              log_(spdlog::level::trace, Eq(absl::string_view("onDownstreamConnectionClose 2 1"))));
+  read_filter_callbacks_.connection_.close(Network::ConnectionCloseType::FlushWrite);
+  // Noop.
+  read_filter_callbacks_.connection_.close(Network::ConnectionCloseType::FlushWrite);
+  filter().testClose();
+}
+
+TEST_P(WasmNetworkFilterTest, RestrictLog) {
+  if (std::get<0>(GetParam()) != "v8" || std::get<1>(GetParam()) != "cpp") {
```
Comment (mathetake) [inline test/extensions/filters/network/wasm/wasm_filter_test.cc:249]: ditto

File: test/extensions/filters/network/wasm/wasm_filter_test.cc (hunk 8 above)
Comment (ryanapilado) [reply test/extensions/filters/network/wasm/wasm_filter_test.cc]: Sorry for the delay. Fixed.

File: bazel/repository_locations.bzl [hunk 10]
```diff
@@ -846,8 +846,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_name = "WebAssembly for Proxies (C++ host implementation)",
         project_desc = "WebAssembly for Proxies (C++ host implementation)",
         project_url = "https://github.com/proxy-wasm/proxy-wasm-cpp-host",
-        version = "6dab125d7a668c7158848b6f48c67fd827c952e6",
-        sha256 = "b5c73ed053a7079bd8bf53b14c4811e87ae521d9fcf4769ec5b248202a27600d",
+        version = "993766996269bd1e41dc625063cfcf4904d940a0",
```
Comment (mathetake) [inline bazel/repository_locations.bzl:849]: Could you update this to the latest? https://github.com/proxy-wasm/proxy-wasm-cpp-host/commit/5a53cf4b231599e1d2a1f2f4598fdfbb727ff948

**Review by mathetake — COMMENTED** [review]
Looks good now. Thanks

File: bazel/repository_locations.bzl (hunk 10 above)
Comment (ryanapilado) [reply bazel/repository_locations.bzl]: Updated

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 11]
```diff
@@ -18,6 +18,26 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
 // [#extension: envoy.bootstrap.wasm]
 
+// Configuration for restricting proxy-wasm capabilities available to modules.
+message CapabilityRestrictionConfig {
+  // The proxy-wasm capabilities which will be allowed. Capabilities are mapped by
+  // name. The *SanitizationConfig* which each capability maps to is currently unimplemented and ignored,
+  // and so should be left empty.
+  //
+  // The capability names are given in the
+  // `proxy-wasm ABI <https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT>`_.
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/wasm.proto:28]: Nit: `s/proxy-wasm/Proxy-Wasm/`

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 12]
```diff
@@ -18,6 +18,26 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#protodoc-title: Wasm]
 // [#extension: envoy.bootstrap.wasm]
 
+// Configuration for restricting proxy-wasm capabilities available to modules.
+message CapabilityRestrictionConfig {
+  // The proxy-wasm capabilities which will be allowed. Capabilities are mapped by
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/wasm.proto:23]: Nit: `s/proxy-wasm/Proxy-Wasm/`

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 13]
```diff
@@ -18,6 +18,26 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  // `proxy-wasm ABI <https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT>`_.
+  // Additionally, the following WASI capabilities from
+  // `this list <https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#modules>`_
+  // are implemented and can be allowed:
+  // *fd_write*, *fd_read*, *fd_seek*, *fd_close*, *fd_fdstat_get*, *environ_get*, *environ_sizes_get*,
+  // *args_get*, *args_sizes_get*, *proc_exit*, *clock_time_get*, *random_get*.
+  map<string, SanitizationConfig> allowed_capabilities = 1;
+}
+
+// Configuration for sanitization of inputs to an allowed capability. NOTE: This is currently unimplemented.
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/wasm.proto:37]: Nit: move `NOTE` to a new line.

File: source/extensions/common/wasm/wasm.cc [hunk 14]
```diff
@@ -312,8 +313,9 @@ WasmEvent toWasmEvent(const std::shared_ptr<WasmHandleBase>& wasm) {
   NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
 }
 
-static bool createWasmInternal(const VmConfig& vm_config, const PluginSharedPtr& plugin,
-                               const Stats::ScopeSharedPtr& scope,
+static bool createWasmInternal(const VmConfig& vm_config,
+                               const CapabilityRestrictionConfig& cr_config,
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm.cc:317]: Nit: `s/cr_config/capabilities_config`, otherwie variable isn't self-describing.

File: test/extensions/common/wasm/test_data/test_restriction_cpp.cc [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+#endif
+
+START_WASM_PLUGIN(CommonWasmRestrictionTestCpp)
+
+WASM_EXPORT(void, proxy_abi_version_0_2_1, (void)) {}
+
+WASM_EXPORT(uint32_t, proxy_on_vm_start, (uint32_t context_id, uint32_t configuration_size)) {
+  (void)(context_id);
+  (void)(configuration_size);
+  std::string level_message = "after on_vm_start, before proxy_log";
```
Comment (PiotrSikora) [inline test/extensions/common/wasm/test_data/test_restriction_cpp.cc:19]: Nit: `s/level_message/log_message/`

Also, this message is written as part of `proxy_log`, not before `proxy_log`.

File: test/extensions/common/wasm/wasm_test.cc [hunk 16]
```diff
@@ -939,6 +953,275 @@ TEST_P(WasmCommonTest, RemoteCodeMultipleRetry) {
...
+
+  EXPECT_FALSE(wasm->capabilityAllowed("proxy_on_vm_start"));
+  EXPECT_FALSE(wasm->capabilityAllowed("proxy_log"));
+  EXPECT_FALSE(wasm->capabilityAllowed("fd_write"));
+
+  EXPECT_NE(wasm, nullptr);
+  auto context = std::make_unique<TestContext>(wasm.get());
+  EXPECT_TRUE(wasm->initialize(code, false));
+
+  // on_vm_start will trigger proxy_log, but expect no call because both are restricted
```
Comment (PiotrSikora) [inline test/extensions/common/wasm/wasm_test.cc:995]: Should `proxy_on_vm_start` callback be even called considering that is not allowed? This is special case, since if we don't call `proxy_on_vm_start` then there is a high chance that plugin won't work correctly, but other callbacks should be skipped if they are not in the allowed capabilities.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 12 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Fixed.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 11 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Fixed.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 13 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Fixed.

File: source/extensions/common/wasm/wasm.cc (hunk 14 above)
Comment (ryanapilado) [reply source/extensions/common/wasm/wasm.cc]: Fixed, changed to capabiility_restriction_config for consistency with type name and proto. Can change to capabilities_config if too long.

File: test/extensions/common/wasm/test_data/test_restriction_cpp.cc (hunk 15 above)
Comment (ryanapilado) [reply test/extensions/common/wasm/test_data/test_restriction_cpp.cc]: Fixed.

File: test/extensions/common/wasm/wasm_test.cc (hunk 16 above)
Comment (ryanapilado) [reply test/extensions/common/wasm/wasm_test.cc]: Yeah, the comment is incorrect, proxy_on_vm_start will never run in this case. Fixed. I also added a different callback to the wasm file (proxy_on_context_create) and adjusted the tests to test for it.

File: api/envoy/extensions/wasm/v3/wasm.proto (hunk 13 above)
Comment (ryanapilado) [reply api/envoy/extensions/wasm/v3/wasm.proto]: Misunderstood this, fixed.

**Review by mathetake — APPROVED** [review]

**Review by PiotrSikora — APPROVED** [review]
Thanks!

@lizan could you take a pass?

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
@envoyproxy/dependency-shepherds PTAL
